src/passes/pass.cpp

```diff
@@ -138,19 +138,23 @@ void PassRunner::addDefaultGlobalOptimizationPasses() {
 }
 
 void PassRunner::run() {
-  if (options.debug) {
+  // BINARYEN_PASS_DEBUG is a convenient commandline way to log out the toplevel passes, their times,
+  //                     and validate between each pass.
+  //                     (we don't recurse pass debug into sub-passes, as it doesn't help anyhow and
+  //                     also is bad for e.g. printing which is a pass)
+  static const int passDebug = getenv("BINARYEN_PASS_DEBUG") ? atoi(getenv("BINARYEN_PASS_DEBUG")) : 0;
+  if (!isNested() && (options.debug || passDebug)) {
```

Comment:
Either specified as another field in the constructor (ew), added to the `PassOptions` struct (that isn't code-smelly but it isn't what that struct is for), or set externally before `run`. So cleanest would probably be:
```c++
PassRunner runner(module);
runner.setIsNested(true);
runner.add<Pass>();
runner.run();
```
Or `PassRunner runner = PassRunner::createNestedPassRunner(module)` that sets the field however it gets set.

And I can't think of a different kind of runner either, hence the `Foo` example. I think I prefer the subclassing idea, just because it's useful for partitioning the nested case away from the non-nested case. Plus we can refactor  later when we have a better idea of what other types of `PassRunner`s exist.
My main objection to a virtual `isNested` method is is that it uses polymorphism to dynamically dispatch the value of a flag, so we can switch on that flag to do different logic. Better would be to either polymorphically dispatch to different logic, *or* switch on a non-polymorphic flag to do logic. This introduces two levels of indirection where we really just need one.

As you said before, we may want to do other stuff based on whether a runner is nested or not. But that's what polymorphism is for, we can make anything that changes into a virtual call, and put the nesting-specific logic in the subclass. On the other hand, if splitting up the logic like that is too inconvenient, then let's not deal with the (mental and design) overhead of having a subclass, and encode nestedness as a member field.

I'm ok with either method, so long as they're consistent.